python/taichi/misc/np2ply.py

```diff
@@ -128,6 +125,19 @@ def add_vertex_rgba(self, r: np.array, g: np.array, b: np.array, a: np.array):
         self.add_vertex_channel("blue", "float", b)
         self.add_vertex_channel("Alpha", "float", a)
 
+    # pass ti vector/matrix tensor directly
+    def add_vertex_rgba(self, rgba):
+        if not isinstance(rgba, np.ndarray):
+            rgba = rgba.to_numpy()
+        channels = rgba.shape[rgba.ndim-1]
+        assert channels == 4, "The dimension for rgba should be 4"
+        n = rgba.size // channels
+        rgba = np.reshape(rgba, (n, channels))

```

Comment:
If we're going to reshape rgba, could we do that at first, at least before we check its shape in line 132?

Reply:
To me I would insist on checking the last dimension first, just to make sure the tensor people passed at least has the same dimension as `rgba` required. Otherwise, if we reshape first i.e. a `100*100*2` array into `5000*4` one, and by chance `vertex_num = 5000` then an unnoticeable runtime bug is there.

Reply:
Well I have different opinions. It doesn't matter we pass 100*100*2 or 5000*4, it should work in both situations, or the reshape loses its meaning.

Reply:
haha I assume this problem is the same as that in the conversation above. For a professional programmer, even the function needs i.e. `vector<vector4D>`, she/he could store it whatever the shape outside, as long as its convenient.

For a rookie programmer like myself I would appreciate less freedom in this input list just to prevent errors T_T. Yeah, I think I will adopt your suggestion here by assuming the programmer has handled the size well.

Reply:
I think in Python there is a unspoken rule that if something will get reshaped eventually, its input dimension doesn't matter any more. We could stick to this convention here

Reply:
> unspoken rule

That frightens me a bit T_T, okay I will leave this conversation unresolved just in case related improvement can be done in the future. For now, I will keep it a little strict on the input requirement just in case for errors.

Reply:
The issue with this approach is that if the user passes something else that has exactly 4 channels, it will only get an error from reshape, and that is very discouraging for a user. For most, people want a transparent yet robust tool, not delving into our tool and dealing with all the unexpected errors. We should at least ensure that our code can handle all the unexpected situations without any uncaptured error message directly from runtime.

Reply:
> ...if the user passes something else that has exactly 4 channels, it will only get an error from reshape...

Although this will be addressed as in the screenshot, I would like to adopt your final suggestion about which api to keep that explicitly indicating all input channels by the user is the most robust.

![image](https://user-images.githubusercontent.com/48085697/84558124-02e2b180-ad63-11ea-8f25-76c487099944.png)

I will comment out those "compact version api" and add a comment note in the code, just in case they will be reactive in the future.

Thanks for this long patience, I learned a lot :>